pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	// timer is used when performing delayed writes to filesystem.
+	timer       *time.Timer
+	timerMutex  sync.Mutex
+	updateDelay time.Duration
+
+	systemContext *ciTypes.SystemContext
+}
+
+var _ Store = (*FileStore)(nil)
+
+// fileStoreOption is used to provide functional options to the fileStore.
+type fileStoreOption func(*FileStore)
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{SystemRegistriesConfPath: s.configPath}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	// Using a different mutex to protect timer vs. rules to avoid deadlock.
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
+	}
+
+	s.timer = time.AfterFunc(s.updateDelay, func() {
+		if err := s.updateConfigNow(); err != nil {
+			log.Errorf("Failed to update the registry mirror config: %v", err)
+		}
+	})
+
+	return nil
+}
+
+func (s *FileStore) updateConfigNow() error {
+	icspRules, idmsRules, itmsRules := s.getAllMirrorSets()
+
+	// Populate config.
+	config := new(sysregistriesv2.V2RegistriesConf)
+	err := registries.EditRegistriesConfig(config, nil, nil, icspRules, idmsRules, itmsRules)
+	if err != nil {
+		return errors.Wrap(err, "could not create registries config")
```

Comment:
nit: this is potentially editing something that already exists, not creating

Reply:
The error itself is in a location where the consolidated `config` object is being created, so I believe the message is accurate. We're not editing the config in place, always 'replacing' it with a new one.